### ISSUE TITLE
Make duplicate check ignore accents/trema in Russian

### ISSUE THREAD

**Issue by Heineken** [ankidroid/Anki-Android#4821]
[x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[x] I have searched for similar existing issues here and on the user forum

###### Reproduction Steps

1. add Russian word to deck
2. try to add the same word again with accent (indicating the stress), e.g. from QuickDic

###### Expected Result
Red highlight as duplicate indication

###### Actual Result
 Duplicate is not recognized

I suggest making the duplicate detection ignore accents/trema on Russian vowels, as
1. the trema is optional ("ё" often written as "е")
2. accents are only used in dictionaries to indicate the stressed syllable

I see the potential hitch with other languages where accents are used and make a difference between two different words (e.g. Spanish "se"≠"sé").
Maybe this can be limited to Russian/Cyrillic letters? (I don't know if there are other Cyrillic alphabets using accents, though.)
Or maybe the difference is that the Spanish "é" is an actual character opposed to a Russian vowel + an accent (separate combining Unicode character)?

Anyway, this is just an idea and not urgent by any means.

**Comment by timrae** [comment]
This sounds pretty complicated and error-prone

**Comment by bkhl** [comment]
Note that in my duplicate ticket #4695 I'm talking about Thai, where the _ideal_ behaviour would be somewhat different. The modifying characters are effectively separate graphemes.

If it's difficult to highlight only the incorrect diacritic/vowel mark/tone mark/&c. maybe a different colour (orange?) could be used to indicate the main character in the box is correct, but the diacritic wrong?

**Comment by mikehardy** [comment]
I see what you mean - could add that into #4695 ? It will serve two purposes - you'll be listening on that issue for the future, and any work in the area will take your use case into account (ideally). Anything happening in type-answer regarding diacritics will likely happen at the same time as it is a very localized feature so it makes sense to still have it one spot I think

**Comment by milyukhin** [comment]
As I understood, searching for duplicates is done by [computing a checksum of a string value](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/Utils.java#L508), so we won't find the de-glyphed fields unless we [persisted](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/Note.java#L151) de-glyphed versions of them in the first place. This leads us to changing `Utils#fieldChecksum(String)`, so the unglyphed version of the field is saved and found later as expected.

Is my understanding correct? If so should we update all the pre-existing in DB checksums when new Anki released? 

Something like this passes:
```java
    @Test
    public void testFileChecksum() {
        long en_e = Utils.fieldChecksum(removeDiacriticalSymbols("e"));
        long es_e_diacritics = Utils.fieldChecksum(removeDiacriticalSymbols("é"));
        long ru_e = Utils.fieldChecksum(removeDiacriticalSymbols("е"));
        long ru_e_diacritics = Utils.fieldChecksum(removeDiacriticalSymbols("ё"));
        long en_a = Utils.fieldChecksum(removeDiacriticalSymbols("a"));
        long es_a_diacritics = Utils.fieldChecksum(removeDiacriticalSymbols("á"));

        assertEquals(en_e, es_e_diacritics);
        assertEquals(ru_e, ru_e_diacritics);
        assertEquals(en_a, es_a_diacritics);
    }

    public static String removeDiacriticalSymbols(String string) {
        return Normalizer.normalize(string, Normalizer.Form.NFD)
                .replaceAll("\\p{InCombiningDiacriticalMarks}+", "");
    }
```

**Comment by Heineken** [comment]
> 
> 
> Note that in my duplicate ticket #4695 I'm talking about Thai, where the _ideal_ behaviour would be somewhat different. The modifying characters are effectively separate graphemes.
> 
> If it's difficult to highlight only the incorrect diacritic/vowel mark/tone mark/&c. maybe a different colour (orange?) could be used to indicate the main character in the box is correct, but the diacritic wrong?

Just to make sure there are no misunderstandings:
If I understand correctly, the Thai issue (#4695) refers to a check performed when typing an answer (didn't even know this feature). This issue, however, refers to the duplicate check performed while creating a note.
Then again, this comment may be irrelevant because the same check is evoked under the hood in both cases anyway.

**Comment by Heineken** [comment]
Thanks for figuring and trying that out, @milyukhin!
If it works as you described and both behaviours (ignoring and respecting diacritics) is intended, then I guess, yeah, both checksum must be stored.
That doubles the storage size for checksums. That is probably no issue, right?
But why do you want to implement storing these before the code using them is implemented?

**Comment by milyukhin** [comment]
That's how I see this:
Currently, when we create a new note and put _å_ as a front text for the note, Anki stores checksum = _3484749270_ for the front.
When I create another note with _a_ as a front Anki calculates checksum of it (which is _2264392759_) and will be searching if it already exists. It doesn't as the checksums are different.

If we want the letters to be treated as same, we need Anki to de-glyphed the letters before calculate the checksum (store the de-glyphed versions of the checksums).
Then on de-glyphed _å_ checksum will be = _2264392759_, and when the new note with front text _a_ is created (same checksum), Anki would be able found the checksum of _å_ and mark it as a duplicate.

What I meant by updating the pre-existed database, is updating what already exists in `notes` table (the checksum column) to be able effectively find duplicates.

Somebody please correct me where I'm wrong

**Comment by Heineken** [comment]
Nothing to correct here, apparently. Can you implement that?

**Comment by adamwolf** [comment]
How does Anki Desktop do this?

**Comment by Heineken** [comment]
Good question!
I've never used it.

**Comment by mikehardy** [comment]
Copied by request from #5611 where @maxily has what looks like an excellent working prototype of comparing normalized text.

TL;DR is I think we need to work to get this upstream if we want it here. I think it has potential, for what that's worth.

---

Very interesting!

This is a requested feature and as a language learner in a language with diacritical marks (but thankfully not so many - Spanish) this would be useful to me as well. Here is my question - how will this interoperate in the Anki ecosystem? 

Their implementation looks like 

```python
def checksum(data):
    if isinstance(data, str):
        data = data.encode("utf-8")
    return sha1(data).hexdigest()
```

To the point - if it doesn't work with how Anki does it then we can't do it, we must play well in the ecosystem and as a matching concern it is difficult enough for us to track anki features (as a constraint imposed by available developer time) that we typically don't support extra features beyond Anki even if we can make them compatible

But if anki can do it as well (maybe a PR there? or some relevant discussion?) then maybe we can do it.

I will say that in general the style would be *how* this would work (a migration, and a Utils alteration). The details might need little tweaks but if this feature happened I think it would follow this sketch

See what they think in Anki upstream and come back here?

**Comment by mikehardy** [comment]
Has anyone brought this up for AnkiDesktop? We need to know how Anki Desktop works before we attempt to do it here

**Comment by maxily** [comment]
Alternatively, we could've introduced a property toggle (say it's off by default - current behavior) which would make AnkiDroid ignoring diacritical marks when it's turned on

**Comment by mikehardy** [comment]
@maxily sure, but this feels like an ecosystem-wide type deal. Wouldn't everyone want this? And the thinking around it will be identical even if the implementation might diverge a bit between the two. I imagine it will be behind a preference for both (some people will likely want to test if they can get their words diacritically correct...). I still want to see what upstream thinks

**Comment by david-allison-1** [comment]
As far as I know, this is now implemented in Anki Desktop

**Comment by abdnh** [comment]
This is not actually implemented in Anki Desktop for duplicate checks, but Anki supports the `nc:` search syntax for ignoring diacritics, which I think is also useful to have in AnkiDroid. (I'm thinking of submitting a PR to add it to AnkiDroid by the way)

**Comment by david-allison-1** [comment]
I've implemented the Rust backend and will be moving to use its search capabilities once we've upgraded the database to schema 16 in AnkiDroid 2.16 (the backend is pinned to Anki 2.135 right now until we match its capabilities).  

If it's a quick change, go for it. If it'll be a lot of effort, it should be included in my changes for AnkiDroid 2.16.

**Comment by abdnh** [comment]
This is great news. Thank you for your hard work!
This should not be a lot of effort, but I'm not familiar with AnkiDroid's codebase, so I'll just give up and wait for your changes if this will cause me headaches 😅
